---
title: "Exploratory analysis and visualization -- HTG"
author: "Hildah Njoroge"
date: "2023-07-17"
output: html_document
---


### Load necessary packages
```{r}
library(DESeq2)
library(pheatmap)
library(ggplot2)
library(ggrepel)
library(RColorBrewer)
library(limma)
```


### Load data

Load the data saved as a DDS object from the HTG and the HTSeq pipelines

### 1. zero and low expression values filtered from the HTSeq pipleline

```{r}

load("C:/hne_files/MSc_Bioinf/MScProject-main_C/MScProject-main_10.06.23/scripts/DEG_analysis/Gene_counts/hisat2_htseq_pipeline/htseq_dds_filtering.Rdata")
ddsHTSeq
```

```{r}
dds_cpm_filtered_HTSeq 
```

```{r}
dds_rowSums_greaterthan_10_HTSeq
```

### versus 
### 2. zero and low expression values filtered from the HTG pipleline
```{r}
load("C:/hne_files/MSc_Bioinf/MScProject-main_C/MScProject-main_10.06.23/scripts/DEG_analysis/Gene_counts/htg_pipeline/htg_dds_filtering.Rdata")
ddsHTG
```

```{r}
dds_cpm_filtered_HTG 
```


```{r}
dds_rowSums_greaterthan_10_HTG
```



## Most statistical methods for EDA of high dimensional data work best of data that is homoskedastic
This poses a challenge with RNA-Seq raw coutn data that is heteroskedastic, i.e., varaince increases (grows) with mmean value
To take care of this, DESEq2 incorporates two methods of data transformation, *rlog - regularised logarithm transformation* and *VST, variance stabilising transformation*. Using ordinary log transformation of normalised data or size-factor normalised data for statistical methods for high dimensional data such as clustering or PCA would not give correct results

### Normalise --> 1. vst for HTG
```{r}
dds <- ddsHTG
normCounts_vst <- vst(dds, blind = TRUE)
assay(normCounts_vst)[1:5, 1:5] ## counts are not whole numbers as they are fractions of gene counts/median of ratios(size factor) --!! VST does not take into account differences in size factors. This may introduce undesired artifacts for size factors greater than 4

```


### Distribution

```{r}
hist(assay(normCounts_vst), main = "HTG Pipeline output when VST normalised",xlim = c(0,20), breaks = seq(0, 20, by = 0.8))
```

```{r}
normCounts_vst <- vst(dds_cpm_filtered_HTG, blind = TRUE)

dim(normCounts_vst)

hist(assay(normCounts_vst), main = "HTG Pipeline output when VST normalised after log2(cpm) > 1 filtering ", xlim = c(0,20), breaks = seq(0, 20, by = 0.8))

# sampleTable$condition <- relevel(sampleTable$condition, ref = "noColitis")
sampleColor <- brewer.pal(4, "Accent")[1:4]
# # sampleColor <- cols <- c( "red","orange", "#0080FF","lightgreen")
names(sampleColor) <- levels(coldata$condition2)
# sampleTable$condition <- factor(sampleTable$condition)
sampleDist_vst <- cor(assay(normCounts_vst), method = "spearman")
pheatmap(sampleDist_vst,
         clustering_distance_rows = as.dist(1 - sampleDist_vst),
         clustering_distance_cols = as.dist(1 - sampleDist_vst),
         annotation_col = data.frame(Disease = coldata$condition2,
                                     row.names = row.names(coldata)),
         annotation_colors = list(Disease = sampleColor),
         main = "Heatmap after filtering by \
         log2(cpm) > 1 & VST normalised approach - Spearman corr")


sampleDist_vst <- cor(assay(normCounts_vst), method = "pearson")
pheatmap(sampleDist_vst,
         clustering_distance_rows = as.dist(1 - sampleDist_vst),
         clustering_distance_cols = as.dist(1 - sampleDist_vst),
         annotation_col = data.frame(Disease = coldata$condition2,
                                     row.names = row.names(coldata)),
         annotation_colors = list(Disease = sampleColor),
         main = "Heatmap after filtering by \
         log2(cpm) > 1 & VST normalised approach - Pearson corr")
```


```{r}
normCounts_vst <- vst(dds_rowSums_greaterthan_10_HTG, blind = TRUE)
dim(normCounts_vst)

hist(assay(normCounts_vst), main = "HTG Pipeline output when VST normalised after raw counts > 10 filtering ", xlim = c(0,20),  breaks = seq(0, 20, by = 0.8))

sampleDist_vst <- cor(assay(normCounts_vst), method = "spearman")
pheatmap(sampleDist_vst,
         clustering_distance_rows = as.dist(1 - sampleDist_vst),
         clustering_distance_cols = as.dist(1 - sampleDist_vst),
         annotation_col = data.frame(Disease = coldata$condition2,
                                     row.names = row.names(coldata)),
         annotation_colors = list(Disease = sampleColor),
         main = "Heatmap after filtering by \
         raw count >10 & VST normalised approach - Spearman corr")


sampleDist_vst <- cor(assay(normCounts_vst), method = "pearson")
pheatmap(sampleDist_vst,
         clustering_distance_rows = as.dist(1 - sampleDist_vst),
         clustering_distance_cols = as.dist(1 - sampleDist_vst),
         annotation_col = data.frame(Disease = coldata$condition2,
                                     row.names = row.names(coldata)),
         annotation_colors = list(Disease = sampleColor),
         main = "Heatmap after filtering by \
         raw count >10 & VST normalised approach - Pearson corr")
```
###  Plotting all the three histograms side by side

```{r}
# par( mfrow = c( 2, 2 ) )
# 
# #Distribution
# hist(assay(normCounts_vst), main = "HTG Pipeline output when VST normalised",xlim = c(0,20),  breaks = seq(0, 20, by = 0.2))
# 
# hist(assay(normCounts_vst), main = "HTG Pipeline output when VST \
# normalised after log2(cpm) > 1 filtering ", xlim = c(0,20), breaks = seq(0, 20, by = 0.2))
# 
# hist(assay(normCounts_vst), main = "HTG Pipeline output when VST \
# normalised after raw counts > 10 filtering ", xlim = c(0,20), breaks = seq(0, 20, by = 0.2))
```

We specify blind=FALSE, which means that differences across donors and treatment should not add to the variance-mean profile of the experiment. However, the experimental design is not used directly in the transformation, only in estimating the global amount of variability in the counts. For a fully unsupervised transformation, one can set blind=TRUE (which is the default) 

```{r}
dds <- ddsHTG
rld <- rlog(dds, blind=FALSE); #head(assay(rld), 3)
assay(rld)[1:5,1:5]
```
 
 
To show the effect of the transformation, in the figure below is a plot of the first sample against the second, first simply using the log2 function (after adding 1, to avoid taking the log of zero), and then using the rlog-transformed values. 
For the log2 approach, we need to first estimate size factors to account for sequencing depth, and then specify normalized=TRUE. 
Sequencing depth correction is done automatically for the rlog method (and for varianceStabilizingTransformation).

```{r}

par( mfrow = c( 1, 2 ) ) 
dds <- estimateSizeFactors(dds) 
plot(log2(counts(dds, normalized=TRUE)[,1:2] + 1), pch=16, cex=0.5, cex.main=0.9, main="Log2 of size-factor normalised gene counts") 
abline(a = 0, b = 1, col = "red")
plot(assay(rld)[,1:2], pch=16, cex=0.5,cex.main=0.9, main=("r-log normalised gene counts"))
# plot(assay(normCounts_vst[,1:2]),pch=16, cex=0.5, main="VST normalised gene counts")
abline(a = 0, b = 1, col = "red")

# par(mfrow = c(1, 2))
# 
# plot(1:10, main = "Plot 1")
# plot(11:20, main = "Plot 2")
# 
# # Add a title to the entire plot
# main_title <- "Side-by-Side Plots"
# mtext(main_title, outer = TRUE, line = 0.5, cex = 1.5)

```

```{r}
par( mfrow = c( 1, 3 ) ) 
dds <- estimateSizeFactors(dds) 
plot(log2(counts(dds, normalized=TRUE)[,1:2] + 1), pch=16, cex=0.7, main="Log2 of size-factor normalised gene counts")
abline(a = 0, b = 1, col = "red")
plot(assay(rld)[,1:2], pch=16, cex=0.7, main=("r-log normalised gene counts"))
abline(a = 0, b = 1, col = "red")
plot(assay(normCounts_vst[,1:2]),pch=16, cex=0.7, main="VST normalised gene counts")
abline(a = 0, b = 1, col = "red")

# Add a title to the entire plot
# main_title <- "Side-by-Side plots of the first sample against\
# the second using different approaches of normalising the gene expression data"
# # mtext(main_title, outer = TRUE, line = 0.5, cex = 1.5)
# mtext(main_title, side = 3, line = 2, cex = 1)
```

#### Distribution of r-log transformed gene counts

```{r}
normCounts_rlog <- rlog(dds, blind = TRUE)
```
```{r}
hist(assay(normCounts_rlog), main = "HTG Pipeline output when rlog normalised",xlim = c(-5,30), breaks = seq(-5, 30, by = 0.8))
```



```{r}
normCounts_rlog <- rlog(dds_cpm_filtered_HTG, blind = TRUE)

```
```{r}

dim(normCounts_rlog)

hist(assay(normCounts_rlog), main = "HTG Pipeline output when r-log normalised after log2(cpm) > 1 filtering ",  xlim = c(-5,30), breaks = seq(-5, 30, by = 0.8))



sampleDist_rlog <- cor(assay(normCounts_rlog), method = "spearman")
# sampleColor <- brewer.pal(4, "Accent")[1:4]
# sampleColor <- cols <- c( "red","orange", "#0080FF","lightgreen")
names(sampleColor) <- levels(coldata$condition2)
pheatmap(sampleDist_rlog,
         clustering_distance_rows = as.dist(1 - sampleDist_rlog),
         clustering_distance_cols = as.dist(1 - sampleDist_rlog),
         annotation_col = data.frame(Disease = coldata$condition2,
                                     row.names = row.names(coldata)),
         annotation_colors = list(Disease = sampleColor),
         main = "Heatmap after filtering by \
         log2(cpm) > 1 & r-log normalised approach - Spearman corr)")


sampleDist_rlog <- cor(assay(normCounts_rlog), method = "pearson")
# sampleColor <- brewer.pal(4, "Accent")[1:4]
# sampleColor <- cols <- c( "red","orange", "#0080FF","lightgreen")
names(sampleColor) <- levels(coldata$condition2)
pheatmap(sampleDist_rlog,
         clustering_distance_rows = as.dist(1 - sampleDist_rlog),
         clustering_distance_cols = as.dist(1 - sampleDist_rlog),
         annotation_col = data.frame(Disease = coldata$condition2,
                                     row.names = row.names(coldata)),
         annotation_colors = list(Disease = sampleColor),
         main = "Heatmap after filtering by \
        log2(cpm) > 1 & r-log normalised approach - Pearson corr)")
```



```{r}
normCounts_rlog <- rlog(dds_rowSums_greaterthan_10_HTG, blind = TRUE)
```

## pearson's correlation is sensitive to outliers

```{r}

dim(normCounts_rlog)

hist(assay(normCounts_rlog), main = "HTG Pipeline output when rlog normalised after raw counts > 10 filtering ", xlim = c(-5,30),  breaks = seq(-5, 30, by = 0.8))


sampleDist_rlog <- cor(assay(normCounts_rlog), method = "spearman")
sampleColor <- brewer.pal(4, "Accent")[1:4]
# sampleColor <- cols <- c( "red","orange", "#0080FF","lightgreen")
names(sampleColor) <- levels(coldata$condition2)
pheatmap(sampleDist_rlog,
         clustering_distance_rows = as.dist(1 - sampleDist_rlog),
         clustering_distance_cols = as.dist(1 - sampleDist_rlog),
         annotation_col = data.frame(Disease = coldata$condition2,
                                     row.names = row.names(coldata)),
         annotation_colors = list(Disease = sampleColor),
         main = "Heatmap after filtering by \
         raw count >10 & r-log normalised approach - Spearman corr)")

sampleDist_rlog <- cor(assay(normCounts_rlog), method = "pearson")

sampleColor <- brewer.pal(4, "Accent")[1:4]
# sampleColor <- cols <- c( "red","orange", "#0080FF","lightgreen")
names(sampleColor) <- levels(coldata$condition2)
pheatmap(sampleDist_rlog,
         clustering_distance_rows = as.dist(1 - sampleDist_rlog),
         clustering_distance_cols = as.dist(1 - sampleDist_rlog),
         annotation_col = data.frame(Disease = coldata$condition2,
                                     row.names = row.names(coldata)),
         annotation_colors = list(Disease = sampleColor),
         main = "Heatmap after filtering by \
         raw count >10 & r-log normalised approach - Pearson corr)")

```



```{r}
# par( mfrow = c( 1, 2 ) )
# 
# #Distribution
# hist(assay(rld), main = "HTG Pipeline output when r-log normalised",xlim = c(0,20),  breaks = seq(0, 20, by = 0.2))
# 
# hist(assay(rld), main = "HTG Pipeline output when r-log \
# normalised after log2(cpm) > 1 filtering ", xlim = c(0,20), breaks = seq(0, 20, by = 0.2))
# # 
# # hist(assay(rld), main = "HTG Pipeline output when r-log \
# # normalised after raw counts > 10 filtering ", xlim = c(0,20), breaks = seq(0, 20, by = 0.2))
```


## Plotting the sample heatmap
#### first, generate the similarity/distance matrix showing the pairwise distances (similarity or dissimilarity between samples).
#### 
```{r}
sampleDist_rlog <- cor(assay(normCounts_rlog), method = "spearman")
sampleDist_rlog[1:6,1:6]
```

```{r}
sampleDist_rlog <- cor(assay(normCounts_rlog), method = "pearson")
sampleDist_rlog[1:6,1:6]
```


## QS: wby use the Spearman option? IS it because the distribution of normalised count data is not normally distributed? Right sqewed? 


```{r}
sampleDist_vst <- cor(assay(normCounts_vst), method = "spearman")
sampleDist_vst[1:6,1:6]
```

```{r}
par(mfrow=c(3,1))

sampleDist_rlog <- cor(assay(normCounts_rlog), method = "spearman")

sampleColor <- brewer.pal(4, "Accent")[1:4]
# sampleColor <- cols <- c( "red","orange", "#0080FF","lightgreen")
names(sampleColor) <- levels(coldata$condition2)
pheatmap(sampleDist_rlog,
         clustering_distance_rows = as.dist(1 - sampleDist_rlog),
         clustering_distance_cols = as.dist(1 - sampleDist_rlog),
         annotation_col = data.frame(Disease = coldata$condition2,
                                     row.names = row.names(coldata)),
         annotation_colors = list(Disease = sampleColor),
         main = "Heatmap after filtering by \
         raw count >10 & r-log normalised approach - Spearman corr)")

sampleDist_rlog <- cor(assay(normCounts_rlog), method = "pearson")

sampleColor <- brewer.pal(4, "Accent")[1:4]
# sampleColor <- cols <- c( "red","orange", "#0080FF","lightgreen")
names(sampleColor) <- levels(coldata$condition2)
pheatmap(sampleDist_rlog,
         clustering_distance_rows = as.dist(1 - sampleDist_rlog),
         clustering_distance_cols = as.dist(1 - sampleDist_rlog),
         annotation_col = data.frame(Disease = coldata$condition2,
                                     row.names = row.names(coldata)),
         annotation_colors = list(Disease = sampleColor),
         main = "Heatmap after filtering by \
         raw count >10 & r-log normalised approach - Pearson corr)")


sampleDist_vst <- cor(assay(normCounts_vst), method = "spearman")
pheatmap(sampleDist_vst,
         clustering_distance_rows = as.dist(1 - sampleDist_vst),
         clustering_distance_cols = as.dist(1 - sampleDist_vst),
         annotation_col = data.frame(Disease = coldata$condition2,
                                     row.names = row.names(coldata)),
         annotation_colors = list(Disease = sampleColor),
         main = "Heatmap after filtering by \
         raw count >10 & VST normalised approach - Spearman corr")

 


```
## this was calculated using the example from the paper *RNA-Seq workflow: Gene-level exploratory analysis and differential expression, by Love et.al *

```{r}

sampleDist_rlog2 <- dist( t( assay(normCounts_rlog) ) ) 
sampleDist_rlog2[1:6]

sampleDistMatrix <- as.matrix( sampleDist_rlog2 ) 
sampleDistMatrix[1:6,1:6]
# rownames(sampleDistMatrix) <- paste( rld$dex, rld$cell, sep="-" ) 
# colnames(sampleDistMatrix) <- NULL
colnames(sampleDistMatrix) <- rownames(coldata)
sampleColor <- brewer.pal(4, "Accent")[1:4]
# sampleColor <- cols <- c( "red","orange", "#0080FF","lightgreen")
names(sampleColor) <- levels(coldata$condition2)
# colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255) 
# pheatmap(sampleDistMatrix,
#          clustering_distance_rows=sampleDist_rlog2, 
#          clustering_distance_cols=sampleDist_rlog2, 
#          col=colors)

pheatmap(sampleDistMatrix,
         clustering_distance_rows=sampleDist_rlog2, 
         clustering_distance_cols=sampleDist_rlog2,
         annotation_col = data.frame(Disease = coldata$condition2,
                                     row.names = row.names(coldata)),
         annotation_colors = list(Disease = sampleColor),
         main = "Heatmap after filtering by \
         raw count >10 & r-log normalised approach (Euclidean Dist)")
```

#### Sample PCA
```{r}

par(mfrow=c(3,1))
pcaRes <- prcomp(t(assay(normCounts_rlog)))
varExp <- round(pcaRes$sdev^2 / sum(pcaRes$sdev^2) * 100)
pcaDF <- data.frame(
  PC1 = pcaRes$x[, 1],
  PC2 = pcaRes$x[, 2],
  Disease = coldata$condition2,
  Sample = row.names(coldata))
pcaPlot <- ggplot(
  data = pcaDF,
  mapping = aes(x = PC1, y = PC2, color = Disease, label = Sample)) +
  geom_point(size = 3) +
  geom_text_repel(size = 2) +
  labs(x = paste0("PC1 (", varExp[1], " %)"), y = paste0("PC2 (", varExp[2], " %)")) +
  theme_minimal() +
  theme(axis.text = element_text(size = 12), legend.text = element_text(size = 10)) +
  scale_color_manual(values = brewer.pal(4, "Accent")[1:4])
print(pcaPlot)

pcaRes <- prcomp(t(assay(normCounts_vst)))
varExp <- round(pcaRes$sdev^2 / sum(pcaRes$sdev^2) * 100)
pcaDF <- data.frame(
  PC1 = pcaRes$x[, 1],
  PC2 = pcaRes$x[, 2],
  Disease = coldata$condition2,
  Sample = row.names(coldata))
pcaPlot <- ggplot(
  data = pcaDF,
  mapping = aes(x = PC1, y = PC2, color = Disease, label = Sample)) +
  geom_point(size = 3) +
  geom_text_repel(size = 0) +
  labs(x = paste0("PC1 (", varExp[1], " %)"), y = paste0("PC2 (", varExp[2], " %)")) +
  theme_minimal() +
  theme(axis.text = element_text(size = 12), legend.text = element_text(size = 10)) +
  scale_color_manual(values = brewer.pal(4, "Accent")[1:4])
print(pcaPlot)
```

#### PCA of the diseased group only

step 1: create the colData for the disease group
```{r}
library(dplyr)

disease_coldata <- coldata %>% 
  filter(condition2 != "noColitis")

```

2. normalise the disease samples and plot the clustering by sample and gene expression 
3. Plot the PCA

```{r}
par(mfrow=c(3,1))
# pcaRes <- prcomp(t(assay(normCounts_rlog)))
# varExp <- round(pcaRes$sdev^2 / sum(pcaRes$sdev^2) * 100)
# pcaDF <- data.frame(
#   PC1 = pcaRes$x[, 1],
#   PC2 = pcaRes$x[, 2],
#   Disease = disease_coldata$condition2,
#   Sample = row.names(coldata))
# pcaPlot <- ggplot(
#   data = pcaDF,
#   mapping = aes(x = PC1, y = PC2, color = Disease, label = Sample)) +
#   geom_point(size = 3) +
#   geom_text_repel(size = 0) +
#   labs(x = paste0("PC1 (", varExp[1], " %)"), y = paste0("PC2 (", varExp[2], " %)")) +
#   theme_minimal() +
#   theme(axis.text = element_text(size = 12), legend.text = element_text(size = 10)) +
#   scale_color_manual(values = brewer.pal(4, "Accent")[1:4])
# print(pcaPlot)

pcaRes <- prcomp(t(assay(normCounts_vst)))
varExp <- round(pcaRes$sdev^2 / sum(pcaRes$sdev^2) * 100)
pcaDF <- data.frame(
  PC1 = pcaRes$x[, 1],
  PC2 = pcaRes$x[, 2],
  Disease = disease_coldata$condition2,
  Sample = row.names(disease_coldata))
pcaPlot <- ggplot(
  data = pcaDF,
  mapping = aes(x = PC1, y = PC2, color = Disease, label = Sample)) +
  geom_point(size = 3) +
  geom_text_repel(size = 4) +
  labs(x = paste0("PC1 (", varExp[1], " %)"), y = paste0("PC2 (", varExp[2], " %)")) +
  theme_minimal() +
  theme(axis.text = element_text(size = 12), legend.text = element_text(size = 10)) +
  scale_color_manual(values = brewer.pal(4, "Accent")[1:4])
print(pcaPlot)
```


## Save the work as an R image

```{r}
save.image(file = "C:/hne_files/MSc_Bioinf/MScProject-main_C/MScProject-main_10.06.23/scripts/DEG_analysis/Gene_counts/exploratory_data_analysis_n_sampleQC.Rdata")

```

